llvm/test/CodeGen/RISCV/memcmp-optsize.ll

```diff
@@ -315,967 +3233,10985 @@ define i32 @bcmp_size_31(ptr %s1, ptr %s2) nounwind optsize {
+; CHECK-UNALIGNED-RV64-V-NEXT:    lui a4, 16
+; CHECK-UNALIGNED-RV64-V-NEXT:    addiw a4, a4, -1
+; CHECK-UNALIGNED-RV64-V-NEXT:    and a2, a2, a4
+; CHECK-UNALIGNED-RV64-V-NEXT:    srli a5, a3, 8
+; CHECK-UNALIGNED-RV64-V-NEXT:    slli a3, a3, 8
+; CHECK-UNALIGNED-RV64-V-NEXT:    or a3, a3, a5
+; CHECK-UNALIGNED-RV64-V-NEXT:    and a3, a3, a4
+; CHECK-UNALIGNED-RV64-V-NEXT:    bne a2, a3, .LBB24_2
+; CHECK-UNALIGNED-RV64-V-NEXT:  # %bb.1: # %loadbb1
+; CHECK-UNALIGNED-RV64-V-NEXT:    lbu a0, 2(a0)
+; CHECK-UNALIGNED-RV64-V-NEXT:    lbu a1, 2(a1)
+; CHECK-UNALIGNED-RV64-V-NEXT:    sub a0, a0, a1
+; CHECK-UNALIGNED-RV64-V-NEXT:    ret
+; CHECK-UNALIGNED-RV64-V-NEXT:  .LBB24_2: # %res_block
+; CHECK-UNALIGNED-RV64-V-NEXT:    sltu a0, a2, a3
+; CHECK-UNALIGNED-RV64-V-NEXT:    neg a0, a0
+; CHECK-UNALIGNED-RV64-V-NEXT:    ori a0, a0, 1
+; CHECK-UNALIGNED-RV64-V-NEXT:    ret
 entry:
-  %bcmp = call signext i32 @bcmp(ptr %s1, ptr %s2, iXLen 63)
-  ret i32 %bcmp
+  %memcmp = call signext i32 @memcmp(ptr %s1, ptr %s2, iXLen 3)
+  ret i32 %memcmp
 }
 
-define i32 @bcmp_size_64(ptr %s1, ptr %s2) nounwind optsize {
-; CHECK-RV32-LABEL: bcmp_size_64:
-; CHECK-RV32:       # %bb.0: # %entry
-; CHECK-RV32-NEXT:    addi sp, sp, -16
-; CHECK-RV32-NEXT:    sw ra, 12(sp) # 4-byte Folded Spill
-; CHECK-RV32-NEXT:    li a2, 64
-; CHECK-RV32-NEXT:    call bcmp
-; CHECK-RV32-NEXT:    lw ra, 12(sp) # 4-byte Folded Reload
-; CHECK-RV32-NEXT:    addi sp, sp, 16
-; CHECK-RV32-NEXT:    ret
+define i32 @memcmp_size_4(ptr %s1, ptr %s2) nounwind optsize {
+; CHECK-ALIGNED-RV32-LABEL: memcmp_size_4:
+; CHECK-ALIGNED-RV32:       # %bb.0: # %entry
+; CHECK-ALIGNED-RV32-NEXT:    lbu a2, 0(a0)
+; CHECK-ALIGNED-RV32-NEXT:    lbu a3, 1(a0)
+; CHECK-ALIGNED-RV32-NEXT:    lbu a4, 2(a0)
+; CHECK-ALIGNED-RV32-NEXT:    lbu a0, 3(a0)
+; CHECK-ALIGNED-RV32-NEXT:    lbu a5, 0(a1)
+; CHECK-ALIGNED-RV32-NEXT:    lbu a6, 1(a1)
+; CHECK-ALIGNED-RV32-NEXT:    lbu a7, 2(a1)
+; CHECK-ALIGNED-RV32-NEXT:    lbu a1, 3(a1)
+; CHECK-ALIGNED-RV32-NEXT:    slli a4, a4, 8
+; CHECK-ALIGNED-RV32-NEXT:    or a0, a4, a0
+; CHECK-ALIGNED-RV32-NEXT:    slli a3, a3, 16
+; CHECK-ALIGNED-RV32-NEXT:    slli a2, a2, 24
+; CHECK-ALIGNED-RV32-NEXT:    or a2, a2, a3
+; CHECK-ALIGNED-RV32-NEXT:    or a0, a2, a0
+; CHECK-ALIGNED-RV32-NEXT:    slli a7, a7, 8
+; CHECK-ALIGNED-RV32-NEXT:    or a1, a7, a1
+; CHECK-ALIGNED-RV32-NEXT:    slli a6, a6, 16
+; CHECK-ALIGNED-RV32-NEXT:    slli a5, a5, 24
+; CHECK-ALIGNED-RV32-NEXT:    or a2, a5, a6
+; CHECK-ALIGNED-RV32-NEXT:    or a1, a2, a1
+; CHECK-ALIGNED-RV32-NEXT:    sltu a2, a1, a0
+; CHECK-ALIGNED-RV32-NEXT:    sltu a0, a0, a1
+; CHECK-ALIGNED-RV32-NEXT:    sub a0, a2, a0
+; CHECK-ALIGNED-RV32-NEXT:    ret
+;
+; CHECK-ALIGNED-RV64-LABEL: memcmp_size_4:
+; CHECK-ALIGNED-RV64:       # %bb.0: # %entry
+; CHECK-ALIGNED-RV64-NEXT:    lbu a2, 0(a0)
+; CHECK-ALIGNED-RV64-NEXT:    lbu a3, 1(a0)
+; CHECK-ALIGNED-RV64-NEXT:    lbu a4, 2(a0)
+; CHECK-ALIGNED-RV64-NEXT:    lb a0, 3(a0)
```

Comment:
Why didn't this use a lbu to avoid the andi later?

Reply:
It seems to be related to https://reviews.llvm.org/D130397. Maybe we should revert this change or add a pattern to catch this.